### ISSUE TITLE
-Wdeclaration-after-statement doesn't work when used with -std=c99

### ISSUE THREAD

**Issue by lukakudra** [llvm/llvm-project#51273]
|  |  |
| --- | --- |
| Bugzilla Link | [51931](https://llvm.org/bz51931) |
| Version | 12.0 |
| OS | Linux |
| CC | @dwblaikie,@zygoloid |

## Extended Description 
-Wdeclaration-after-statement doesn't do anything if combined with -std=c99 or newer.

Take a look at the following program:

// prog.c
#include <stdio.h>

int main(void)
{
    printf("hello world\n");
    int i = 0;

    return 0;
}

If I compile it with clang with the following command:

$ clang -std=c99 -Wdeclaration-after-statement prog.c

it produces no warnings.

If I compile the same code with gcc with the following command:

$ gcc -std=c99 -Wdeclaration-after-statement prog.c

it produces the following warning:

prog.c: In function ‘main’:
prog.c:6:9: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
    6 |         int i = 0;
      |         ^~~

This is the behavior I would like to have with clang, but it only produces this warning if I use it with -std=c90 or -std=c89 or -ansi, like this:

$ clang -std=c90 -Wdeclaration-after-statement prog.c
prog.c:6:6: warning: ISO C90 forbids mixing declarations and code [-Wdeclaration-after-statement]
        int i = 0;
            ^
1 warning generated.

I asked this question on stackoverflow because I thought I was using the flags in a wrong way but it turns that -Wdeclaration-after-statement just doesn't work with clang in the scenario that I'm using it.
You can take a look at the question on the following link:
https://stackoverflow.com/questions/69270600/how-to-force-werror-declaration-after-statement-with-std-c99-in-clang

To summarize -Wdeclaration-after-statement doesn't work with -std=c99 on clang, but it works on gcc.

I tested this on Arch Linux with clang version 12.0.1.

**Comment by zero9178** [comment]
Potential fix in https://reviews.llvm.org/D114787

**Comment by zero9178** [comment]
Fixed in https://github.com/llvm/llvm-project/commit/118f966b46cfb60897b56a9878e1c68fd0e2afa4
